L1Trigger/L1THGCal/src/concentrator/HGCalConcentratorSuperTriggerCellImpl.cc

```diff
@@ -86,53 +123,310 @@ HGCalConcentratorSuperTriggerCellImpl::getSuperTriggerCellId(int detid) const {
+                math::PtEtaPhiMLorentzVector p4(0, point.eta(), point.phi(), 0.);
+                
+                triggerCell.setPosition(point);     
+                triggerCell.setP4(p4);
+                triggerCell.setDetId(newtc);
+
+                s.second.addToList( triggerCell );                      
+                trigCellVecOutput.push_back ( triggerCell );
+                
+              }
+              
+            }
+            
+          }
+          
+        }
+        
+      }
+     
+
+ 
+    }
+
+        
+ 
+}
+
+
+void 
+HGCalConcentratorSuperTriggerCellImpl::
+coarsenTriggerCells( std::unordered_map<unsigned,SuperTriggerCell>& STCs, const std::vector<l1t::HGCalTriggerCell>& trigCellVecInput, std::vector<l1t::HGCalTriggerCell>& trigCellVecOutput)
+{ 
+  std::unordered_map<unsigned,SuperTriggerCell> coarseTCs;
+
+
+  for (const l1t::HGCalTriggerCell & tc : trigCellVecInput) {
+    if (tc.subdetId() != HGCHEB) {
+      coarseTCs[getSuperTriggerCellId(tc.detId(),2)].add(tc);
+    }
+    else{
+      trigCellVecOutput.push_back( tc );
+    }
+  }
+
+  for (const l1t::HGCalTriggerCell & tc : trigCellVecInput) {
+     if (tc.subdetId() == HGCHEB) continue;
+
+     int thickness = triggerTools_.thicknessIndex(tc.detId(),true);
+     if ( thickness > 0 ){
+
+      if (!(tc.detId() & 1)) { //if even
+        const auto & ctc = coarseTCs[getSuperTriggerCellId(tc.detId(),2)]; 
+        auto & stc = STCs[getSuperTriggerCellId(tc.detId())]; 
+        trigCellVecOutput.push_back( tc );
+        ctc.assignEnergy(trigCellVecOutput.back(), "STC");      
+        //        //reassign max id to the correct one;
+        if ( trigCellVecOutput.back().hwPt() >= stc.GetMaxHwPt() ){
+
+          if ( stc.GetMaxId() != tc.detId() ){
+            stc.SetMaxTC( trigCellVecOutput.back() );
+          }
+
+        }
+      }
+
+    }
+    else{
+
+      trigCellVecOutput.push_back( tc );
+
+    }
+
+
+  }
+
+
+}
+
+
 void 
 HGCalConcentratorSuperTriggerCellImpl::
 superTriggerCellSelectImpl(const std::vector<l1t::HGCalTriggerCell>& trigCellVecInput, std::vector<l1t::HGCalTriggerCell>& trigCellVecOutput)
 { 
 
+  fixedDataSize_ = true;
+
   std::unordered_map<unsigned,SuperTriggerCell> STCs;
+  std::vector<l1t::HGCalTriggerCell> trigCellVecInputEnlarged;
 
-  // first pass, fill the super trigger cells
+  // first pass, fill the "coarse" trigger cells
   for (const l1t::HGCalTriggerCell & tc : trigCellVecInput) {
+    trigCellVecInputEnlarged.push_back ( tc );
     if (tc.subdetId() == HGCHEB) continue;
     STCs[getSuperTriggerCellId(tc.detId())].add(tc);
   }
-    
+
+
+  //The missing trigger cells are needed to be created both for coarsening the existing TC
+  // in the thick modules (for the fixed data size choice), or for any of the energy spread algorithms (except super TCs)
+
+  if ( fixedDataSize_ == true || energyDivisionType_!=superTriggerCell){
+    createMissingTriggerCells( STCs, trigCellVecInputEnlarged);
```

Comment:
It would be simpler (and clearer) to create **all** TCs from STCs and not just "missing" TCs. And special treatment of "super TC" vs "coarse TC" wouldn't be needed.

Reply:
So recreate an existing TC potentially? This could be simpler, but is it additional overhead?

Reply:
Anyway all TCs have to be recreated (either with a copy or from scratch). The input collection is constant and cannot be touched, so all the output objects need to be constructed.

Reply:
The problem I am currently having is in the `coarsenTriggerCells` function
In this case we need to remove TCs from the list if they are odd, and if they are in the thick layers.
In the simplest case I can have the input vector (containing all TCs including the newly created ones) and an output vector where I copy the TCs if they are even, or in the thin layer.
But this might be expensive. I have tried looking at `remove_if` and lambda operators, but I couldn't determine if this is what we needed - do you have any suggestions?

Reply:
I don't think it is a good idea to implement the coarsening of trigger cells here. 
If we need to have coarser trigger cells, the ideal solution is to have coarser TC definition in TC construction. But a better temporary solution would be to have an other implementation class that does this coarsening and called before STC.

Reply:
I agree with you about the ideal solution.
But for now we would have the following issues:
1) The "even" TCs would have to be created in the coarsening implementation class, so duplicating some of the `createMissingTriggerCells` function.
2) The TCs which were removed by the coarsening implementation would then be recreated in the `createMissingTriggerCells` function of HGCalConcentratorSuperTriggerCellImpl.cc
Also we only want to coarsen cells in the low-density layers.
Maybe there's a way around this, that you can see?

Reply:
With the latest commit I basically managed to do it, although there is some duplication of code.
The TCs are firstly coarsened in their own implementation class (at this stage they are "1x2 superTCs" rather than coarse-2).
Then in the SuperTriggerCellImpl I modify the position of those "1x2 superTCs" to be the average of the odd and even cells.
The code compiles, but I would have to run some validation that the logic is correct. Let me know what you think.